lib/elixir/lib/kernel/typespec.ex

```diff
@@ -212,8 +212,9 @@ defmodule Kernel.Typespec do
     :ets.take(bag, key)
   end
 
-  defp translate_type({_, {kind, {:::, _, [{name, _, args}, definition]}, pos}})
+  defp translate_type({_, {kind, expr = {:::, _, [{name, _, args}, definition]}, pos}})
        when is_atom(name) and name != ::: do
+    validate_type_ast!(expr)
```

Comment:
Since we are only interested in validating arguments, I think we can name it more explicitly and pass less data downstream:
```
validate_type_args!(type_name, arguments)
```